external_drivers/dhi_dfs/mdal_dhi_dfs2.cpp

```diff
@@ -0,0 +1,240 @@
+      LPCTSTR projectionWkt;
+      if ( geoInfoType == F_UTM_PROJECTION )
+      {
+        double lon, lat, orientation;
+        error = dfsGetGeoInfoUTMProj( pdfs, &projectionWkt, &lon, &lat, &orientation );
+
+        if ( error == NO_ERROR )
+        {
+          mesh->mWktProjection = projectionWkt;
+        }
+
+        System::String ^stringProjection = gcnew System::String( projectionWkt );
+        MapProjection ^mapProj = gcnew MapProjection( stringProjection, false );
+        double N, E;
+        mapProj->Geo2Proj( lon, lat, E, N );
+        double projOri = mapProj->Geo2ProjRotation( lon, lat, orientation );
+
+        // projOri: clockwise rotation of grid
+        // I direction = X direction if projOri=0;
+        // J direction = Y direction if projOri=0;
+        double dIx = sizeX * cos( projOri * M_PI / 180 ); //offset of X coordinate for each element in I direction
+        double dIy = -sizeX * sin( projOri * M_PI / 180 ); //offset of Y coordinate for each element in I direction
+        double dJx = sizeY * sin( projOri * M_PI / 180 ); //offset of X coordinate for each element in J direction
+        double dJy = sizeY * cos( projOri * M_PI / 180 ); //offset of Y coordinate for each element in J direction
+
+        double vertexX0;
+        double vertexY0;
+        if ( projectionWkt == "NON-UTM" )
+        {
+          vertexX0 = E + originX;
+          vertexY0 = N + originY;
+        }
+        else
+        {
+          vertexX0 = E + originX - ( dIx + dJx ) / 2;
+          vertexY0 = N + originY - ( dIy + dJy ) / 2;
+        }
+
+        mesh->buildMesh( vertexX0, vertexY0, dIx, dIy, dJx, dJy, countX, countY );
+
+        ok = mesh->populateDatasetGroups();
+      }
+    }
+  }
+
+  if ( ok )
+    return mesh;
+  else
+  {
+    dfsFileClose( pdfs, &Fp );
+    return nullptr;
+  }
+}
+
+void MeshDfs2::buildMesh( double vertexX0, double vertexY0, double dIx, double dIy, double dJx, double dJy, size_t countI, size_t countJ )
+{
+  // build vertices
+  mVertexCoordinates = std::vector<double>( ( countI + 1 ) * ( countJ + 1 ) * 3, 0 );
+  for ( size_t j = 0; j <= countJ; ++j )
+    for ( size_t i = 0; i <= countI; ++i )
+    {
+      mVertexCoordinates[3 * ( i + j * ( countI + 1 ) )] = vertexX0 + dIx * i + dJx * j;
+      mVertexCoordinates[3 * ( i + j * ( countI + 1 ) ) + 1] = vertexY0 + dIy * i + dJy * j;
+    }
+
+  std::vector<double> xBounding( 4, 0 );
+  std::vector<double> yBounding( 4, 0 );
+
+  xBounding[0] = mVertexCoordinates.at( 0 );
+  yBounding[0] = mVertexCoordinates.at( 1 );
+  xBounding[1] = mVertexCoordinates.at( 3 * countI );
+  yBounding[1] = mVertexCoordinates.at( 3 * countI + 1 );
+  xBounding[2] = mVertexCoordinates.at( 3 * ( countI + 1 ) * ( countJ ) );
+  yBounding[2] = mVertexCoordinates.at( 3 * ( countI + 1 ) * ( countJ ) + 1 );
+  xBounding[3] = mVertexCoordinates.at( mVertexCoordinates.size() - 3 );
+  yBounding[3] = mVertexCoordinates.at( mVertexCoordinates.size() - 2 );
```

Comment:
sure -3 for x and -2 for y is OK?

Reply:
yes sure, because we have `mVertexCoordinates = std::vector<double>( ( countI + 1 ) * ( countJ + 1 ) * 3, 0 );` with `countI>=0` and `countJ>=0`